Deck 1.5.3  clicking on timeline shows a spinner that doesnt stop

<!--
Thanks for reporting issues back!

Guidelines for submitting issues:

* Please search the existing issues first, it's likely that your issue was already reported or even fixed.
  
* SECURITY: Report any potential security bug to us via our HackerOne page (https://hackerone.com/nextcloud) following our security policy (https://nextcloud.com/security/) instead of filing an issue in our bug tracker.  

* The issues in other components should be reported in their respective repositories: You will find them in our GitHub Organization (https://github.com/nextcloud/)
  
* You can also use the Issue Template app to prefill most of the required information: https://apps.nextcloud.com/apps/issuetemplate
-->

<!--- Please keep this note for other contributors -->

**Describe the bug**
When clicking on timeline for a deck card a spinner comes up and never stops.


**To Reproduce**
Steps to reproduce the behavior:
1. Go to a deck card or complete board
2. Click on the card and click on timeline
3.  See a spinner that never ends.

**Expected behavior**
A clear view of the timeline should be there and the spinner should stop when it has loaded the timeline

**Screenshots**
![Screenshot from 2021-10-29 12-31-39](https://user-images.githubusercontent.com/317830/139420345-e0b8b77d-4d61-4e71-9786-5563a077b878.png)


**Client details:**
 - OS: Ubuntu 
 - Browser chrome and firefox
 - Version [e.g. 22]
 - Device: desktop
<details>
<summary>Server details</summary>

**Operating system**:
Debian 10

**Web server:**
Apache and nginx

**Database:**
mariadb

**PHP version:**
php8 and php7.4

**Nextcloud version:** (see Nextcloud admin page)
22.1.1 and 22.2.0

**Where did you install Nextcloud from:**
Source

**Signing status:**

```
No errors
```

**List of activated apps:**

```
Enabled:
  - accessibility: 1.7.0
  - activity: 2.15.0
  - admin_audit: 1.11.0
  - analytics: 3.6.1
  - appointments: 1.10.2
  - approval: 1.0.7
  - bruteforcesettings: 2.2.0
  - calendar: 2.3.3
  - circles: 22.1.1
  - cloud_federation_api: 1.4.0
  - collectives: 0.14.22
  - comments: 1.11.0
  - contacts: 4.0.3
  - contactsinteraction: 1.2.0
  - dashboard: 7.1.0
  - dav: 1.18.0
  - deck: 1.5.3
  - extract: 1.3.2
  - federatedfilesharing: 1.11.0
  - federation: 1.11.0
  - files: 1.16.0
  - files_accesscontrol: 1.12.0
  - files_automatedtagging: 1.12.0
  - files_downloadactivity: 1.11.1
  - files_external: 1.12.1
  - files_fulltextsearch: 22.0.1
  - files_markdown: 2.3.4
  - files_pdfviewer: 2.3.0
  - files_rightclick: 1.1.0
  - files_sharing: 1.13.2
  - files_trashbin: 1.11.0
  - files_versions: 1.14.0
  - files_videoplayer: 1.11.0
  - flow_notifications: 1.2.0
  - forms: 2.3.0
  - fulltextsearch: 22.0.1
  - fulltextsearch_elasticsearch: 22.0.1
  - groupfolders: 10.0.0
  - guests: 2.0.2
  - login_notes: 0.4.0
  - logreader: 2.7.0
  - lookup_server_connector: 1.9.0
  - mail: 1.10.4
  - nextcloud_announcements: 1.11.0
  - notifications: 2.10.1
  - oauth2: 1.9.0
  - onlyoffice: 7.1.2
  - password_policy: 1.12.0
  - photos: 1.4.0
  - privacy: 1.6.0
  - provisioning_api: 1.11.0
  - recommendations: 1.1.0
  - serverinfo: 1.12.0
  - settings: 1.3.0
  - sharebymail: 1.11.0
  - spreed: 12.1.2
  - support: 1.5.0
  - suspicious_login: 4.0.0
  - systemtags: 1.11.0
  - tasks: 0.14.2
  - text: 3.3.0
  - theming: 1.12.0
  - twofactor_admin: 3.1.0
  - twofactor_backupcodes: 1.10.1
  - twofactor_totp: 6.1.0
  - twofactor_u2f: 6.2.0
  - updatenotification: 1.11.0
  - user_status: 1.1.1
  - viewer: 1.6.0
  - weather_status: 1.1.0
  - welcome: 1.0.0
  - workflow_pdf_converter: 1.7.0
  - workflow_script: 1.7.0
  - workflowengine: 2.3.1

```

**Nextcloud configuration:**

```
{
    "system": {
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "***REMOVED SENSITIVE VALUE***"
        ],
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "dbtype": "mysql",
        "version": "22.1.1.2",
        "overwrite.cli.url": "***REMOVED SENSITIVE VALUE***",
        "overwriteprotocol": "https",
        "htaccess.RewriteBase": "\/",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbport": "",
        "dbtableprefix": "oc_",
        "mysql.utf8mb4": true,
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "installed": true,
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "redis": {
            "host": "***REMOVED SENSITIVE VALUE***",
            "port": 0,
            "timeout": 0,
            "password": "***REMOVED SENSITIVE VALUE***"
        },
        "memcache.local": "\\OC\\Memcache\\Redis",
        "memcache.locking": "\\OC\\Memcache\\Redis",
        "filelocking.enabled": true,
        "enable_previews": true,
        "enabledPreviewProviders": [
            "OC\\Preview\\PNG",
            "OC\\Preview\\JPEG",
            "OC\\Preview\\GIF",
            "OC\\Preview\\BMP",
            "OC\\Preview\\XBitmap",
            "OC\\Preview\\MarkDown",
            "OC\\Preview\\MP3",
            "OC\\Preview\\TXT",
            "OC\\Preview\\Illustrator",
            "OC\\Preview\\Movie",
            "OC\\Preview\\MSOffice2003",
            "OC\\Preview\\MSOffice2007",
            "OC\\Preview\\MSOfficeDoc",
            "OC\\Preview\\OpenDocument",
            "OC\\Preview\\PDF",
            "OC\\Preview\\Photoshop",
            "OC\\Preview\\Postscript",
            "OC\\Preview\\StarOffice",
            "OC\\Preview\\SVG",
            "OC\\Preview\\TIFF",
            "OC\\Preview\\Font"
        ],
        "preview_max_x": 1024,
        "preview_max_y": 768,
        "preview_max_scale_facto": 1,
        "auth.bruteforce.protection.enabled": true,
        "trashbin_retention_obligation": "auto,7",
        "skeletondirectory": "\/data\/nc-skeleton",
        "defaultapp": "file",
        "activity_expire_days": 14,
        "integrity.check.disabled": false,
        "updater.release.channel": "stable",
        "theme": "",
        "default_phone_region": "SE",
        "maintenance": false,
        "app_install_overwrite": [
            "groupfolders"
        ],
        "mail_from_address": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpmode": "smtp",
        "mail_sendmailmode": "smtp",
        "mail_domain": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpauthtype": "LOGIN",
        "mail_smtpauth": 1,
        "mail_smtphost": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpport": "587",
        "mail_smtpname": "***REMOVED SENSITIVE VALUE***",
        "mail_smtppassword": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpsecure": "tls",
        "twofactor_enforced": "true",
        "twofactor_enforced_groups": [],
        "twofactor_enforced_excluded_groups": [],
        "loglevel": 4
    }
}

```

</details>

<details>
<summary>Logs</summary>

#### Nextcloud log (data/nextcloud.log)
No related log to find 

#### Browser log
```
main.js:63 Error: Request failed with status code 304
    at t.exports (createError.js:16)
    at t.exports (settle.js:17)
    at XMLHttpRequest.A.onreadystatechange (xhr.js:62)
```

</details>

--- a/lib/Dashboard/DeckWidgetToday.php
+++ b/lib/Dashboard/DeckWidgetToday.php
@@ -80,6 +80,6 @@ class DeckWidgetToday implements IWidget {
 	 * @inheritDoc
 	 */
 	public function load(): void {
-		\OCP\Util::addScript('deck', 'dashboard');
+		\OCP\Util::addScript('deck', 'deck-dashboard');
 	}
 }

--- a/lib/Dashboard/DeckWidgetTomorrow.php
+++ b/lib/Dashboard/DeckWidgetTomorrow.php
@@ -80,6 +80,6 @@ class DeckWidgetTomorrow implements IWidget {
 	 * @inheritDoc
 	 */
 	public function load(): void {
-		\OCP\Util::addScript('deck', 'dashboard');
+		\OCP\Util::addScript('deck', 'deck-dashboard');
 	}
 }

--- a/src/App.vue
+++ b/src/App.vue
@@ -194,4 +194,10 @@ export default {
 		width: 100%;
 	}
 
+	.modal__card {
+		height: 100vh;
+		max-height: calc(100vh - 120px);
+		overflow: auto;
+	}
+
 </style>

--- a/src/components/ActivityEntry.vue
+++ b/src/components/ActivityEntry.vue
@@ -25,7 +25,7 @@
 		<div class="activity--header">
 			<img :src="activity.icon" class="activity--icon">
 			<NcRichText class="activity--subject" :text="message.subject" :arguments="message.parameters" />
-			<div class="activity--timestamp">
+			<div class="activity--timestamp" :title="formatReadableDate(activity.datetime)">
 				{{ relativeDate(activity.datetime) }}
 			</div>
 		</div>
@@ -39,6 +39,7 @@ import { NcRichText, NcUserBubble } from '@nextcloud/vue'
 import moment from '@nextcloud/moment'
 import DOMPurify from 'dompurify'
 import relativeDate from '../mixins/relativeDate.js'
+import formatReadableDate from '../mixins/readableDate.js'
 
 const InternalLink = {
 	name: 'InternalLink',
@@ -62,7 +63,7 @@ export default {
 	components: {
 		NcRichText,
 	},
-	mixins: [relativeDate],
+	mixins: [relativeDate, formatReadableDate],
 	props: {
 		activity: {
 			type: Object,

--- a/src/components/ActivityList.vue
+++ b/src/components/ActivityList.vue
@@ -84,7 +84,20 @@ export default {
 			params.append('object_id', '' + this.objectId)
 			params.append('limit', ACTIVITY_FETCH_LIMIT)
 
-			const response = await axios.get(generateOcsUrl(`apps/activity/api/v2/activity/${this.filter}`) + '?' + params)
+			const response = await axios.get(
+				generateOcsUrl(`apps/activity/api/v2/activity/${this.filter}`) + '?' + params,
+				{
+					validateStatus: (status) => {
+						return (status >= 200 && status < 300) || status === 304
+					},
+				},
+			)
+
+			if (response.status === 304) {
+				this.endReached = true
+				return []
+			}
+
 			let activities = response.data.ocs.data
 			if (this.filter === 'deck') {
 				// We need to manually filter activities here, since currently we use two different types and there is no way
@@ -95,7 +108,7 @@ export default {
 				})
 			}
 			this.activities.push(...activities)
-			if (response.data.ocs.meta.statuscode === 304 || activities.length === 0) {
+			if (activities.length === 0) {
 				this.endReached = true
 				return []
 			}

--- a/src/components/board/Stack.vue
+++ b/src/components/board/Stack.vue
@@ -217,7 +217,7 @@ export default {
 			})
 		},
 		dragHandleSelector() {
-			return this.canEdit ? null : '.no-drag'
+			return this.canEdit && !this.showArchived ? null : '.no-drag'
 		},
 		cardDetailsInModal: {
 			get() {

--- a/src/components/board/TagsTabSidebar.vue
+++ b/src/components/board/TagsTabSidebar.vue
@@ -130,7 +130,7 @@ export default {
 			return true
 		},
 		labelsSorted() {
-			return [...this.labels].sort((a, b) => (a.title < b.title) ? -1 : 1)
+			return [...this.labels].sort((a, b) => a.title.localeCompare(b.title))
 		},
 
 	},

--- a/src/components/card/CardSidebar.vue
+++ b/src/components/card/CardSidebar.vue
@@ -39,13 +39,7 @@
 				{{ t('deck', 'Open in bigger view') }}
 			</NcActionButton>
 
-			<NcActionButton v-for="action in cardActions"
-				:key="action.label"
-				:close-after-click="true"
-				:icon="action.icon"
-				@click="action.callback(cardRichObject)">
-				{{ action.label }}
-			</NcActionButton>
+			<CardMenuEntries :card="currentCard" :hide-details-entry="true" />
 		</template>
 
 		<NcAppSidebarTab id="details"
@@ -104,6 +98,7 @@ import ActivityIcon from 'vue-material-design-icons/LightningBolt.vue'
 
 import { showError } from '@nextcloud/dialogs'
 import { getLocale } from '@nextcloud/l10n'
+import CardMenuEntries from '../cards/CardMenuEntries.vue'
 
 const capabilities = window.OC.getCapabilities()
 
@@ -121,6 +116,7 @@ export default {
 		AttachmentIcon,
 		CommentIcon,
 		HomeIcon,
+		CardMenuEntries,
 	},
 	mixins: [relativeDate],
 	props: {

--- a/src/components/card/CardSidebar.vue
+++ b/src/components/card/CardSidebar.vue
@@ -235,7 +235,7 @@ section.app-sidebar__tab--active {
 	right: 0;
 	max-width: calc(100% - #{$modal-padding * 2});
 	padding: 0 14px;
-	max-height: 100%;
+	height: auto;
 	overflow: initial;
 	user-select: text;
 	-webkit-user-select: text;

--- a/src/components/card/CardSidebarTabDetails.vue
+++ b/src/components/card/CardSidebarTabDetails.vue
@@ -119,8 +119,10 @@ export default {
 		this.initialize()
 	},
 	methods: {
-		descriptionChanged(newDesc) {
-			this.$store.dispatch('updateCardDesc', { ...this.card, description: newDesc })
+		async descriptionChanged(newDesc) {
+			if (newDesc === this.copiedCard.description) {
+				return
+			}
 			this.copiedCard.description = newDesc
 		},
 		async initialize() {

--- a/src/components/card/Description.vue
+++ b/src/components/card/Description.vue
@@ -187,17 +187,26 @@ export default {
 	mounted() {
 		this.setupEditor()
 	},
-	beforeDestroy() {
-		this?.editor?.destroy()
+	async beforeDestroy() {
+		await this.destroyEditor()
 	},
 	methods: {
 		async setupEditor() {
-			this?.editor?.destroy()
+			await this.destroyEditor()
+			this.descriptionLastEdit = 0
+			this.description = this.card.description
 			this.editor = await window.OCA.Text.createEditor({
 				el: this.$refs.editor,
 				content: this.card.description,
 				readOnly: !this.canEdit,
+				onLoaded: () => {
+					this.descriptionLastEdit = 0
+				},
 				onUpdate: ({ markdown }) => {
+					if (this.description === markdown) {
+						this.descriptionLastEdit = 0
+						return
+					}
 					this.description = markdown
 					this.updateDescription()
 				},
@@ -206,6 +215,10 @@ export default {
 				},
 			})
 
+		},
+		async destroyEditor() {
+			await this.saveDescription()
+			this?.editor?.destroy()
 		},
 		addKeyListeners() {
 			this.$refs.markdownEditor.easymde.codemirror.on('keydown', (a, b) => {
@@ -287,6 +300,7 @@ export default {
 				return
 			}
 			this.descriptionSaving = true
+			await this.$store.dispatch('updateCardDesc', { ...this.card, description: this.description })
 			this.$emit('change', this.description)
 			this.descriptionLastEdit = 0
 			this.descriptionSaving = false

--- a/src/components/cards/CardItem.vue
+++ b/src/components/cards/CardItem.vue
@@ -44,9 +44,8 @@
 					tabindex="0"
 					class="editable"
 					:aria-label="t('deck', 'Edit card title')"
-					@click.stop="startEditing(card)"
 					@keydown.enter.stop.prevent="startEditing(card)">
-					{{ card.title }}
+					<span @click.stop="startEditing(card)">{{ card.title }}</span>
 				</h3>
 				<form v-else-if="editing"
 					v-click-outside="cancelEdit"
@@ -299,7 +298,9 @@ export default {
 				padding-left: 4px;
 				align-self: center;
 				&.editable {
-					cursor: text;
+					span {
+						cursor: text;
+					}
 
 					&:focus {
 						outline: 2px solid var(--color-border-dark);

--- a/src/components/cards/CardMenu.vue
+++ b/src/components/cards/CardMenu.vue
@@ -21,214 +21,24 @@
   -->
 
 <template>
-	<div v-if="card" class="card-menu">
-		<div @click.stop.prevent>
-			<NcActions>
-				<NcActionButton v-if="showArchived === false && !isCurrentUserAssigned"
-					icon="icon-user"
-					:close-after-click="true"
-					@click="assignCardToMe()">
-					{{ t('deck', 'Assign to me') }}
-				</NcActionButton>
-				<NcActionButton v-if="showArchived === false && isCurrentUserAssigned"
-					icon="icon-user"
-					:close-after-click="true"
-					@click="unassignCardFromMe()">
-					{{ t('deck', 'Unassign myself') }}
-				</NcActionButton>
-				<NcActionButton icon="icon-checkmark" :close-after-click="true" @click="changeCardDoneStatus()">
-					{{ card.done ? t('deck', 'Mark as not done') : t('deck', 'Mark as done') }}
-				</NcActionButton>
-				<NcActionButton icon="icon-external" :close-after-click="true" @click="modalShow=true">
-					{{ t('deck', 'Move card') }}
-				</NcActionButton>
-				<NcActionButton v-for="action in cardActions"
-					:key="action.label"
-					:close-after-click="true"
-					:icon="action.icon"
-					@click="action.callback(cardRichObject)">
-					{{ action.label }}
-				</NcActionButton>
-				<NcActionButton icon="icon-settings-dark" :close-after-click="true" @click="openCard">
-					<CardBulletedIcon slot="icon" :size="20" decorative />
-					{{ t('deck', 'Card details') }}
-				</NcActionButton>
-				<NcActionButton :close-after-click="true" @click="archiveUnarchiveCard()">
-					<template #icon>
-						<ArchiveIcon :size="20" decorative />
-					</template>
-					{{ card.archived ? t('deck', 'Unarchive card') : t('deck', 'Archive card') }}
-				</NcActionButton>
-				<NcActionButton v-if="showArchived === false"
-					icon="icon-delete"
-					:close-after-click="true"
-					@click="deleteCard()">
-					{{ t('deck', 'Delete card') }}
-				</NcActionButton>
-			</NcActions>
-		</div>
-		<NcModal v-if="modalShow" :title="t('deck', 'Move card to another board')" @close="modalShow=false">
-			<div class="modal__content">
-				<h3>{{ t('deck', 'Move card to another board') }}</h3>
-				<NcMultiselect v-model="selectedBoard"
-					:placeholder="t('deck', 'Select a board')"
-					:options="activeBoards"
-					:max-height="100"
-					label="title"
-					@select="loadStacksFromBoard" />
-				<NcMultiselect v-model="selectedStack"
-					:placeholder="t('deck', 'Select a list')"
-					:options="stacksFromBoard"
-					:max-height="100"
-					label="title">
-					<span slot="noOptions">
-						{{ t('deck', 'List is empty') }}
-					</span>
-				</NcMultiselect>
-
-				<button :disabled="!isBoardAndStackChoosen" class="primary" @click="moveCard">
-					{{ t('deck', 'Move card') }}
-				</button>
-				<button @click="modalShow=false">
-					{{ t('deck', 'Cancel') }}
-				</button>
-			</div>
-		</NcModal>
+	<div v-if="card" class="card-menu" @click.stop.prevent>
+		<NcActions>
+			<CardMenuEntries :card="card" />
+		</NcActions>
 	</div>
 </template>
 <script>
-import { NcModal, NcActions, NcActionButton, NcMultiselect } from '@nextcloud/vue'
-import { mapGetters, mapState } from 'vuex'
-import ArchiveIcon from 'vue-material-design-icons/Archive.vue'
-import CardBulletedIcon from 'vue-material-design-icons/CardBulleted.vue'
-import axios from '@nextcloud/axios'
-import { generateUrl } from '@nextcloud/router'
-import { getCurrentUser } from '@nextcloud/auth'
-import { showUndo } from '@nextcloud/dialogs'
-
-import '@nextcloud/dialogs/dist/index.css'
+import { NcActions } from '@nextcloud/vue'
+import CardMenuEntries from './CardMenuEntries.vue'
 
 export default {
 	name: 'CardMenu',
-	components: { NcActions, NcActionButton, NcModal, NcMultiselect, ArchiveIcon, CardBulletedIcon },
+	components: { NcActions, CardMenuEntries },
 	props: {
 		card: {
 			type: Object,
 			default: null,
 		},
 	},
-	data() {
-		return {
-			modalShow: false,
-			selectedBoard: '',
-			selectedStack: '',
-			stacksFromBoard: [],
-		}
-	},
-	computed: {
-		...mapGetters([
-			'isArchived',
-			'boards',
-			'cardActions',
-		]),
-		...mapState({
-			showArchived: state => state.showArchived,
-			currentBoard: state => state.currentBoard,
-		}),
-		canEdit() {
-			if (this.currentBoard) {
-				return this.$store.getters.canEdit
-			}
-			const board = this.$store.getters.boards.find((item) => item.id === this.card.boardId)
-			return !!board?.permissions?.PERMISSION_EDIT
-		},
-		isBoardAndStackChoosen() {
-			if (this.selectedBoard === '' || this.selectedStack === '') {
-				return false
-			}
-			return true
-		},
-		isCurrentUserAssigned() {
-			return this.card.assignedUsers.find((item) => item.type === 0 && item.participant.uid === getCurrentUser()?.uid)
-		},
-		activeBoards() {
-			return this.$store.getters.boards.filter((item) => item.deletedAt === 0 && item.archived === false)
-		},
-
-		boardId() {
-			return this.card?.boardId ? this.card.boardId : this.$route.params.id
-		},
-	},
-	methods: {
-		openCard() {
-			const boardId = this.card?.boardId ? this.card.boardId : this.$route.params.id
-			this.$router.push({ name: 'card', params: { id: boardId, cardId: this.card.id } }).catch(() => {})
-		},
-		deleteCard() {
-			this.$store.dispatch('deleteCard', this.card)
-			showUndo(t('deck', 'Card deleted'), () => this.$store.dispatch('cardUndoDelete', this.card))
-		},
-		changeCardDoneStatus() {
-			this.$store.dispatch('changeCardDoneStatus', { ...this.card, done: !this.card.done })
-		},
-		archiveUnarchiveCard() {
-			this.$store.dispatch('archiveUnarchiveCard', { ...this.card, archived: !this.card.archived })
-		},
-		assignCardToMe() {
-			this.$store.dispatch('assignCardToUser', {
-				card: this.card,
-				assignee: {
-					userId: getCurrentUser()?.uid,
-					type: 0,
-				},
-			})
-		},
-		unassignCardFromMe() {
-			this.$store.dispatch('removeUserFromCard', {
-				card: this.card,
-				assignee: {
-					userId: getCurrentUser()?.uid,
-					type: 0,
-				},
-			})
-		},
-		async moveCard() {
-			this.copiedCard = Object.assign({}, this.card)
-			this.copiedCard.stackId = this.selectedStack.id
-			this.$store.dispatch('moveCard', this.copiedCard)
-			if (parseInt(this.boardId) === parseInt(this.selectedStack.boardId)) {
-				await this.$store.commit('addNewCard', { ...this.copiedCard })
-			}
-			this.modalShow = false
-		},
-		async loadStacksFromBoard(board) {
-			try {
-				const url = generateUrl('/apps/deck/stacks/' + board.id)
-				const response = await axios.get(url)
-				this.stacksFromBoard = response.data
-			} catch (err) {
-				return err
-			}
-		},
-	},
 }
 </script>
-
-<style lang="scss" scoped>
-	.modal__content {
-		width: 25vw;
-		min-width: 250px;
-		min-height: 120px;
-		text-align: center;
-		margin: 20px 20px 100px 20px;
-
-		.multiselect {
-			margin-bottom: 10px;
-		}
-	}
-
-	.modal__content button {
-		float: right;
-		margin-top: 50px;
-	}
-</style>

--- a/src/components/cards/CardMenu.vue
+++ b/src/components/cards/CardMenu.vue
@@ -42,6 +42,13 @@
 				<NcActionButton icon="icon-external" :close-after-click="true" @click="modalShow=true">
 					{{ t('deck', 'Move card') }}
 				</NcActionButton>
+				<NcActionButton v-for="action in cardActions"
+					:key="action.label"
+					:close-after-click="true"
+					:icon="action.icon"
+					@click="action.callback(cardRichObject)">
+					{{ action.label }}
+				</NcActionButton>
 				<NcActionButton icon="icon-settings-dark" :close-after-click="true" @click="openCard">
 					<CardBulletedIcon slot="icon" :size="20" decorative />
 					{{ t('deck', 'Card details') }}
@@ -122,6 +129,7 @@ export default {
 		...mapGetters([
 			'isArchived',
 			'boards',
+			'cardActions',
 		]),
 		...mapState({
 			showArchived: state => state.showArchived,

--- a/src/components/cards/CardMenuEntries.vue
+++ b/src/components/cards/CardMenuEntries.vue
@@ -26,22 +26,28 @@
 			<CardBulletedIcon slot="icon" :size="20" decorative />
 			{{ t('deck', 'Card details') }}
 		</NcActionButton>
-		<NcActionButton v-if="showArchived === false && !isCurrentUserAssigned"
+		<NcActionButton v-if="canEdit && !isCurrentUserAssigned"
 			icon="icon-user"
 			:close-after-click="true"
 			@click="assignCardToMe()">
 			{{ t('deck', 'Assign to me') }}
 		</NcActionButton>
-		<NcActionButton v-if="showArchived === false && isCurrentUserAssigned"
+		<NcActionButton v-if="canEdit && isCurrentUserAssigned"
 			icon="icon-user"
 			:close-after-click="true"
 			@click="unassignCardFromMe()">
 			{{ t('deck', 'Unassign myself') }}
 		</NcActionButton>
-		<NcActionButton icon="icon-checkmark" :close-after-click="true" @click="changeCardDoneStatus()">
+		<NcActionButton v-if="canEdit"
+			icon="icon-checkmark"
+			:close-after-click="true"
+			@click="changeCardDoneStatus()">
 			{{ card.done ? t('deck', 'Mark as not done') : t('deck', 'Mark as done') }}
 		</NcActionButton>
-		<NcActionButton icon="icon-external" :close-after-click="true" @click="modalShow=true">
+		<NcActionButton v-if="canEdit"
+			icon="icon-external"
+			:close-after-click="true"
+			@click="modalShow=true">
 			{{ t('deck', 'Move card') }}
 		</NcActionButton>
 		<NcActionButton v-for="action in cardActions"
@@ -51,13 +57,13 @@
 			@click="action.callback(cardRichObject)">
 			{{ action.label }}
 		</NcActionButton>
-		<NcActionButton :close-after-click="true" @click="archiveUnarchiveCard()">
+		<NcActionButton v-if="canEditBoard" :close-after-click="true" @click="archiveUnarchiveCard()">
 			<template #icon>
 				<ArchiveIcon :size="20" decorative />
 			</template>
 			{{ card.archived ? t('deck', 'Unarchive card') : t('deck', 'Archive card') }}
 		</NcActionButton>
-		<NcActionButton v-if="showArchived === false"
+		<NcActionButton v-if="canEdit"
 			icon="icon-delete"
 			:close-after-click="true"
 			@click="deleteCard()">
@@ -136,6 +142,9 @@ export default {
 			currentBoard: state => state.currentBoard,
 		}),
 		canEdit() {
+			return !this.card.archived
+		},
+		canEditBoard() {
 			if (this.currentBoard) {
 				return this.$store.getters.canEdit
 			}

--- a/src/components/cards/CardMenuEntries.vue
+++ b/src/components/cards/CardMenuEntries.vue
@@ -1,0 +1,234 @@
+<!--
+  - @copyright Copyright (c) 2018 Julius Härtl <jus@bitgrid.net>
+  -
+  - @author Julius Härtl <jus@bitgrid.net>
+  -
+  - @license GNU AGPL version 3 or any later version
+  -
+  - This program is free software: you can redistribute it and/or modify
+  - it under the terms of the GNU Affero General Public License as
+  - published by the Free Software Foundation, either version 3 of the
+  - License, or (at your option) any later version.
+  -
+  - This program is distributed in the hope that it will be useful,
+  - but WITHOUT ANY WARRANTY; without even the implied warranty of
+  - MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  - GNU Affero General Public License for more details.
+  -
+  - You should have received a copy of the GNU Affero General Public License
+  - along with this program. If not, see <http://www.gnu.org/licenses/>.
+  -
+  -->
+
+<template>
+	<div>
+		<NcActionButton v-if="!hideDetailsEntry" :close-after-click="true" @click="openCard">
+			<CardBulletedIcon slot="icon" :size="20" decorative />
+			{{ t('deck', 'Card details') }}
+		</NcActionButton>
+		<NcActionButton v-if="showArchived === false && !isCurrentUserAssigned"
+			icon="icon-user"
+			:close-after-click="true"
+			@click="assignCardToMe()">
+			{{ t('deck', 'Assign to me') }}
+		</NcActionButton>
+		<NcActionButton v-if="showArchived === false && isCurrentUserAssigned"
+			icon="icon-user"
+			:close-after-click="true"
+			@click="unassignCardFromMe()">
+			{{ t('deck', 'Unassign myself') }}
+		</NcActionButton>
+		<NcActionButton icon="icon-checkmark" :close-after-click="true" @click="changeCardDoneStatus()">
+			{{ card.done ? t('deck', 'Mark as not done') : t('deck', 'Mark as done') }}
+		</NcActionButton>
+		<NcActionButton icon="icon-external" :close-after-click="true" @click="modalShow=true">
+			{{ t('deck', 'Move card') }}
+		</NcActionButton>
+		<NcActionButton v-for="action in cardActions"
+			:key="action.label"
+			:close-after-click="true"
+			:icon="action.icon"
+			@click="action.callback(cardRichObject)">
+			{{ action.label }}
+		</NcActionButton>
+		<NcActionButton :close-after-click="true" @click="archiveUnarchiveCard()">
+			<template #icon>
+				<ArchiveIcon :size="20" decorative />
+			</template>
+			{{ card.archived ? t('deck', 'Unarchive card') : t('deck', 'Archive card') }}
+		</NcActionButton>
+		<NcActionButton v-if="showArchived === false"
+			icon="icon-delete"
+			:close-after-click="true"
+			@click="deleteCard()">
+			{{ t('deck', 'Delete card') }}
+		</NcActionButton>
+		<NcModal v-if="modalShow" :title="t('deck', 'Move card to another board')" @close="modalShow=false">
+			<div class="modal__content">
+				<h3>{{ t('deck', 'Move card to another board') }}</h3>
+				<NcMultiselect v-model="selectedBoard"
+					:placeholder="t('deck', 'Select a board')"
+					:options="activeBoards"
+					:max-height="100"
+					label="title"
+					@select="loadStacksFromBoard" />
+				<NcMultiselect v-model="selectedStack"
+					:placeholder="t('deck', 'Select a list')"
+					:options="stacksFromBoard"
+					:max-height="100"
+					label="title">
+					<span slot="noOptions">
+						{{ t('deck', 'List is empty') }}
+					</span>
+				</NcMultiselect>
+
+				<button :disabled="!isBoardAndStackChoosen" class="primary" @click="moveCard">
+					{{ t('deck', 'Move card') }}
+				</button>
+				<button @click="modalShow=false">
+					{{ t('deck', 'Cancel') }}
+				</button>
+			</div>
+		</NcModal>
+	</div>
+</template>
+<script>
+import { NcModal, NcActionButton, NcMultiselect } from '@nextcloud/vue'
+import { mapGetters, mapState } from 'vuex'
+import ArchiveIcon from 'vue-material-design-icons/Archive.vue'
+import CardBulletedIcon from 'vue-material-design-icons/CardBulleted.vue'
+import axios from '@nextcloud/axios'
+import { generateUrl } from '@nextcloud/router'
+import { getCurrentUser } from '@nextcloud/auth'
+import { showUndo } from '@nextcloud/dialogs'
+
+import '@nextcloud/dialogs/dist/index.css'
+
+export default {
+	name: 'CardMenuEntries',
+	components: { NcActionButton, NcModal, NcMultiselect, ArchiveIcon, CardBulletedIcon },
+	props: {
+		card: {
+			type: Object,
+			default: null,
+		},
+		hideDetailsEntry: {
+			type: Boolean,
+			default: false,
+		},
+	},
+	data() {
+		return {
+			modalShow: false,
+			selectedBoard: '',
+			selectedStack: '',
+			stacksFromBoard: [],
+		}
+	},
+	computed: {
+		...mapGetters([
+			'isArchived',
+			'boards',
+			'cardActions',
+		]),
+		...mapState({
+			showArchived: state => state.showArchived,
+			currentBoard: state => state.currentBoard,
+		}),
+		canEdit() {
+			if (this.currentBoard) {
+				return this.$store.getters.canEdit
+			}
+			const board = this.$store.getters.boards.find((item) => item.id === this.card.boardId)
+			return !!board?.permissions?.PERMISSION_EDIT
+		},
+		isBoardAndStackChoosen() {
+			if (this.selectedBoard === '' || this.selectedStack === '') {
+				return false
+			}
+			return true
+		},
+		isCurrentUserAssigned() {
+			return this.card.assignedUsers.find((item) => item.type === 0 && item.participant.uid === getCurrentUser()?.uid)
+		},
+		activeBoards() {
+			return this.$store.getters.boards.filter((item) => item.deletedAt === 0 && item.archived === false)
+		},
+
+		boardId() {
+			return this.card?.boardId ? this.card.boardId : this.$route.params.id
+		},
+	},
+	methods: {
+		openCard() {
+			const boardId = this.card?.boardId ? this.card.boardId : this.$route.params.id
+			this.$router.push({ name: 'card', params: { id: boardId, cardId: this.card.id } }).catch(() => {})
+		},
+		deleteCard() {
+			this.$store.dispatch('deleteCard', this.card)
+			showUndo(t('deck', 'Card deleted'), () => this.$store.dispatch('cardUndoDelete', this.card))
+		},
+		changeCardDoneStatus() {
+			this.$store.dispatch('changeCardDoneStatus', { ...this.card, done: !this.card.done })
+		},
+		archiveUnarchiveCard() {
+			this.$store.dispatch('archiveUnarchiveCard', { ...this.card, archived: !this.card.archived })
+		},
+		assignCardToMe() {
+			this.$store.dispatch('assignCardToUser', {
+				card: this.card,
+				assignee: {
+					userId: getCurrentUser()?.uid,
+					type: 0,
+				},
+			})
+		},
+		unassignCardFromMe() {
+			this.$store.dispatch('removeUserFromCard', {
+				card: this.card,
+				assignee: {
+					userId: getCurrentUser()?.uid,
+					type: 0,
+				},
+			})
+		},
+		async moveCard() {
+			this.copiedCard = Object.assign({}, this.card)
+			this.copiedCard.stackId = this.selectedStack.id
+			this.$store.dispatch('moveCard', this.copiedCard)
+			if (parseInt(this.boardId) === parseInt(this.selectedStack.boardId)) {
+				await this.$store.commit('addNewCard', { ...this.copiedCard })
+			}
+			this.modalShow = false
+		},
+		async loadStacksFromBoard(board) {
+			try {
+				const url = generateUrl('/apps/deck/stacks/' + board.id)
+				const response = await axios.get(url)
+				this.stacksFromBoard = response.data
+			} catch (err) {
+				return err
+			}
+		},
+	},
+}
+</script>
+
+<style lang="scss" scoped>
+	.modal__content {
+		width: 25vw;
+		min-width: 250px;
+		min-height: 120px;
+		text-align: center;
+		margin: 20px 20px 100px 20px;
+
+		.multiselect {
+			margin-bottom: 10px;
+		}
+	}
+
+	.modal__content button {
+		float: right;
+		margin-top: 50px;
+	}
+</style>

